src/parser/serializer.ts

```diff
@@ -608,18 +620,39 @@ function serializeDBRef(
 export function serializeInto(
   buffer: Uint8Array,
   object: Document,
-  checkKeys = false,
-  startingIndex = 0,
-  depth = 0,
-  serializeFunctions = false,
-  ignoreUndefined = true,
-  path: Document[] = []
+  checkKeys: boolean,
+  startingIndex: number,
+  depth: number,
+  serializeFunctions: boolean,
+  ignoreUndefined: boolean,
+  path: Set<Document> | null
 ): number {
-  startingIndex = startingIndex || 0;
-  path = path || [];
+  if (path == null) {
+    // We are at the root input
+    if (object == null) {
+      // ONLY the root should turn into an empty document
+      // BSON Empty document has a size of 5 (LE)
+      buffer[0] = 0x05;
+      buffer[1] = 0x00;
+      buffer[2] = 0x00;
+      buffer[3] = 0x00;
+      // All documents end with null terminator
+      buffer[4] = 0x00;
+      return 5;
+    }
+
+    if (Array.isArray(object)) {
+      throw new BSONError('serialize does not support an array as the root input');
+    }
+    if (typeof object !== 'object') {
```

Comment:
@nbbeeken I was curious since this line didn't seem it would catch all expectations in my opinion and added some tests to pass a `Long`, `Int32`, etc and it does indeed serialize them as our custom BSON types will pass this check. So it felt odd to me that `BSON.serialize(1)` would error but `BSON.serialize(new Int32('1'))` would not. This ticket explicitly states to error on "non object inputs" but I just felt the above example unexpected.

Since this is the last ticket in the v5 epic I thought (against or normal process in this special case) that maybe we decide what to do about that here? It just feels incomplete that we might still roundtrip some single values as a result of this change, which might be confusing to users. Thoughts?

Reply:
If I add one more check after this lin (probably a more elegant way) then I do get my expectation of all single values not getting round tripped:

```
    if ('_bsontype' in object) {
      throw new BSONError('serialize does not support non-object as the root input');
    }
```

Reply:
Added a check for `_bsontype` and that it is a string, this aligns with our typescript definition in the driver for not recursing a type: [here](https://github.com/mongodb/node-mongodb-native/blob/main/src/mongo_types.ts#L520)

Which in looking at that type, should we also be looking to prevent those other object types that we consider bson value from being passed as the root? Currently both `Date` and `RegExp` result in an empty document, Buffer will turn into (3 bytes in size) `{ '0': 0, '1': 0, '2': 0, parent: {}, offset: 0 }`. Thoughts?